### ISSUE TITLE
Vendored deps problem with dune 2.2.0

### ISSUE THREAD

**Issue by talex5** [ocaml/dune#3139]
In earlier versions of dune, `-p` ignored vendored libraries. However, this was changed in #3075, breaking e.g. the ocaml-ci build:

 (from https://ci.ocamllabs.io/github/ocurrent/ocaml-ci/commit/d3fdf492175ac0647afa71cde20e51f500aad492/variant/4.08):

```
#=== ERROR while compiling ocaml-ci-client.~dev ===============================#
# context              2.0.6 | linux/x86_64 | ocaml-base-compiler.4.08.1 | pinned(file:///src)
# path                 ~/.opam/4.08/.opam-switch/build/ocaml-ci-client.~dev
# command              ~/.opam/4.08/bin/dune build -p ocaml-ci-client -j 71
# exit-code            1
# env-file             ~/.opam/log/ocaml-ci-client-9-bb845d.env
# output-file          ~/.opam/log/ocaml-ci-client-9-bb845d.out
### output ###
# Error: Conflict between the following libraries:
# - "current_rpc" in _build/default/ocurrent/lib_rpc
# - "current_rpc" in /home/opam/.opam/4.08/lib/current_rpc
#   -> required by library "ocaml-ci-api" in
#      /home/opam/.opam/4.08/lib/ocaml-ci-api
# -> required by executable main in client/dune:4
```

What I (think I) want is:

- When I clone the project and build with dune, it builds everything in parallel, including the vendored packages. It does not run the vendored tests or complain about deprecation errors.
- When I install the project with `-p`, we rely on opam to build and install the vendored package, using the vendored package's opam file. This means that it is only built once and there is no conflict.

Instead, with dune 2.2.0, each package builds its own copy of the vendored library.

Is there a way to get the old behaviour? We have `(lang dune 1.11)` in the `dune-project`.

One possibility is using opam pins, but it's quite messy compared to just updating a git submodule (see https://github.com/ocurrent/ocaml-ci/pull/108).

**Comment by rgrinberg** [comment]
In dune's terms, it seems like you aren't "vendoring" anything. You're just using submodules for package management and would like dune pretend the submodules don't exist in release mode.

Is there a way to perhaps tell opam not to recursively clone the submodules when pinning instead? Because in that case, you want to use opam for package management rather than submodules.

**Comment by talex5** [comment]
I don't really see how the new dune design is supposed to work. Packages within a dune project usually depend on each other, and if there are any vendored libraries then won't this always lead to conflicts like the one above?

**Comment by rgrinberg** [comment]
You can still use vendored packages to build private libraries and binaries. I believe the intention with vendoring was always that you want to include 3rd party packages as part of your distribution. I understand that our message was a bit mixed and confused on this however.

Just to understand your use case better:

* There's no intention to distribute any packages within other packages. When you publish your packags, all your tarballs are maximally "unbundled" and all dependencies are obtained from opam.

* In development, instead of fetching your dependencies from opam you would like to use dune's composability + submodules as a "poor man's" package manager.

Essentially, you just want a way to tell dune that some source directories only contains dependencies for development and they should not be assumed to be a part of your project.

**Comment by diml** [comment]
We just had a chat with Rudi about this. The new behaviour seems right to us; if the sources of a dependency are physically present and vendored in the workspace, then surely Dune should prefer them to what's installed on the system. The filtering done by `-p` is meant for multi-package projects, not for filtering out vendored dependencies. It was a bit of an accident that it did.

However, we agree that it's unfortunate that we broke your workflow with this change. So we are planning to add an option to restore the previous behaviour, i.e. hide vendored directories in release builds.

For the future, we are also thinking that filtering per-package at the level of individual stanzas is not always ideal and we are thinking of introducing a mode where each package will be assigned a directory. Anything under this directory would be part of the said package. That would simplify things a bit and we could start filtering per directory rather than per package, which would be simpler and more flexible.

**Comment by diml** [comment]
@talex5 actually, we just realised that there is a very simple workaround. You can change the opam command to do `rm -rf vendor` before the call to dune. That will give you exactly the behaviour you want and will give us a bit more time to figure out exactly what vendoring workflow we want to support in Dune.

**Comment by diml** [comment]
Does that work for you?

**Comment by talex5** [comment]
The way we have been using it is during development, when we want to depend on a feature in another library that hasn't been released yet. Users build using `dune` (and get the vendored version). The CI builds using opam, installing the vendored packages itself.

For releases, we don't have any vendored dependencies because we release the upstream libraries first.

The main advantage is that updating to some new commit in a vendored library is just `git pull` in its subdirectory and then a `git commit` in the parent, and the dependency is visible to the CI. The other option is to use opam `pin-depends` (e.g. see https://github.com/ocurrent/ocaml-ci/pull/108/commits/556fc0e7a87bc848c9f559ced3aae1826cf16835) but that's very tedious and requires users to mess around with opam and install libraries globally, instead of having dune just pick them up.

The reason for telling dune these are "vendored" is just so dune doesn't try to run the tests or complain about deprecations.

Another option would be to change the CI to build directly using dune (no opam), but at the moment I think it's still worth testing that the opam files work, even if they are generated by dune.

The `rm` thing could work, indeed...

**Comment by diml** [comment]
Thanks for the details. It's useful to know how these things are used in practice.

However, I don't completely follow how this work. Just to make sure I understand correctly:
- while you develop or your machine, you build against the vendored code
- when you release in opam, the vendored code is not part of the release tarball and so dependencies are installed by opam the normal way
- when you dev pin the package with opam, opam pulls in the vendored code and since dune 2.2.0, dune builds against the vendored libraries rather than the pre-installed libraries
- in your CI, you build using opam, which dev-pins the package

So before 2.2.0, in you CI you were building against the released version of your dependencies while since 2.2.0 you are building against the vendored ones.

If this is correct, then when you depend on a feature in another library that hasn't been released yet, then your CI should fail when using dune < 2.2.0. Do you agree?

**Comment by talex5** [comment]
> So before 2.2.0, in you CI you were building against the released version of your dependencies while since 2.2.0 you are building against the vendored ones.

No, the CI pins the vendored libraries using opam and builds against those.

**Comment by diml** [comment]
Just curious, how does that work? i.e. how does the CI determines what dependencies should be pinned and at which revision?

**Comment by talex5** [comment]
I've updated the CI so that it tests projects directly, using `dune build @install @runtest`, rather than using `opam install -t .`. This does mean that people doing multi-project builds with vendored dependencies won't be warned that their project can't be installed using opam. Maybe dune could warn about that itself?

**Comment by talex5** [comment]
(I thought I replied to your previous comment, but GitHub isn't showing it. The answer is that the CI lets opam choose the version, which it will do based on the directory name, if it contains a number. It pins all opam files it finds, excluding `test*` directories.)

**Comment by diml** [comment]
Thanks for the info. Running dune directly seems good :+1:
